linkerd/distribute/src/service.rs

```diff
@@ -0,0 +1,317 @@
+use super::{Distribution, WeightedKeys};
+use indexmap::IndexMap;
+use linkerd_stack::Service;
+use rand::{
+    distributions::{Distribution as _, WeightedError},
+    rngs::SmallRng,
+    SeedableRng,
+};
+use std::{
+    hash::Hash,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+/// A service that distributes requests over a set of backends.
+#[derive(Debug)]
+pub struct Distribute<K, S> {
+    backends: IndexMap<K, S>,
+    selection: Selection<K>,
+
+    /// Stores the index of the backend that has been polled to ready. The
+    /// service at this index will be used on the next invocation of
+    /// `Service::call`.
+    ready_idx: Option<usize>,
+}
+
+/// Holds per-distribution state for a [`Distribute`] service.
+#[derive(Debug)]
+enum Selection<K> {
+    Empty,
+    FirstAvailable,
+    RandomAvailable {
+        keys: Arc<WeightedKeys<K>>,
+        rng: SmallRng,
+    },
+}
+
+// === impl Distribute ===
+
+impl<K: Hash + Eq, S> Distribute<K, S> {
+    pub(crate) fn new(backends: IndexMap<K, S>, dist: Distribution<K>) -> Self {
+        Self {
+            backends,
+            selection: dist.into(),
+            ready_idx: None,
+        }
+    }
+}
+
+impl<Req, K, S> Service<Req> for Distribute<K, S>
+where
+    K: Hash + Eq,
+    S: Service<Req>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    /// Acquires a ready backend.
+    ///
+    /// Note that this doesn't necessarily drive all backend services to
+    /// readiness. We expect that these inner services should be buffered or
+    /// otherwise drive themselves to readiness (i.e. via SpawnReady).
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // If we've already chosen a ready index, then skip polling.
+        if self.ready_idx.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        match self.selection {
+            Selection::Empty => {
+                tracing::debug!("empty distribution will never become ready");
+            }
+
+            Selection::FirstAvailable => {
+                for (idx, svc) in self.backends.values_mut().enumerate() {
+                    if svc.poll_ready(cx)?.is_ready() {
+                        self.ready_idx = Some(idx);
+                        return Poll::Ready(Ok(()));
+                    }
+                }
+            }
+
+            // Choose a random index (via the weighted distribution) to try to
+            // poll the backend. Continue selecting endpoints until we find one
+            // that is ready or we've tried all backends in the distribution.
+            Selection::RandomAvailable {
+                ref keys,
+                ref mut rng,
+            } => {
+                // Clone the weighted index so that we can zero out the weights
+                // as pending services are polled.
+                let mut index = keys.index().clone();
```

Comment:
i wonder if we could reduce allocations slightly by cloning the `WeightedIndex` once per `Distribute` instance and resetting the zeroed weights every time we select a ready index, rather than cloning it once per `poll_ready`? this may be a premature optimization but it could be worth considering at some point...

Reply:
https://github.com/linkerd/linkerd2/issues/9892

Reply:
Yeah, it's complicated enough that I don't think it's worth it presently. We're not expecting this list to be large in the vast majority of situations. Clarity/correctness is far more important than shaving off a small allocation.